### PR TITLE
cli: support executing "initial SQL" from a directory

### LINKED ISSUE #48647: build: docker image should support init scripts in docker-entrypoint-initdb.d
The Postgres (and many other databases') docker image supports adding initialization scripts to `docker-entrypoint-initdb.d`. We should have something similar, since it seems like a nice convention.

---

The Postgres docker docs: https://hub.docker.com/_/postgres

If you would like to do additional initialization in an image derived from this one, add one or more *.sql, *.sql.gz, or *.sh scripts under /docker-entrypoint-initdb.d (creating the directory if necessary). After the entrypoint calls initdb to create the default postgres user and database, it will run any *.sql files, run any executable *.sh scripts, and source any non-executable *.sh scripts found in that directory to do further initialization before starting the service.

Warning: scripts in /docker-entrypoint-initdb.d are only run if you start the container with a data directory that is empty; any pre-existing database will be left untouched on container startup. One common problem is that if one of your /docker-entrypoint-initdb.d scripts fails (which will cause the entrypoint script to exit) and your orchestrator restarts the container with the already initialized data directory, it will not continue on with your scripts.

For example, to add an additional user and database, add the following to /docker-entrypoint-initdb.d/init-user-db.sh:
```
#!/bin/bash
set -e

psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
    CREATE USER docker;
    CREATE DATABASE docker;
    GRANT ALL PRIVILEGES ON DATABASE docker TO docker;
EOSQL
```
These initialization files will be executed in sorted name order as defined by the current locale, which defaults to en_US.utf8. Any *.sql files will be executed by POSTGRES_USER, which defaults to the postgres superuser. It is recommended that any psql commands that are run inside of a *.sh script be executed as POSTGRES_USER by using the --username "$POSTGRES_USER" flag. This user will be able to connect without a password due to the presence of trust authentication for Unix socket connections made inside the container.

Additionally, as of docker-library/postgres#253, these initialization scripts are run as the postgres user (or as the "semi-arbitrary user" specified with the --user flag to docker run; see the section titled "Arbitrary --user Notes" for more details). Also, as of docker-library/postgres#440, the temporary daemon started for these initialization scripts listens only on the Unix socket, so any psql usage should drop the hostname portion (see docker-library/postgres#474 (comment) for example).

Epic CRDB-1424

### PR DESCRIPTION
**Pull request by knz**
**tldr** with this patch, CockroachDB runs `.sql` files from the directory specified via `--initial-sql-dir` exactly once (only on 1 node) upon cluster initialization.

Example use cases for this mechanism:

- create some initial users/roles.
- define some initial passwords.
- create some initial databases/tables.
- load an initial dataset using IMPORT or RESTORE.
- set up cluster settings.

Fixes  #48647. 

- first commit from #54758 - can be ignored.
- 2nd commit (loopback conns) for review by @tbg  
- 3rd commit (exactly-once init) for review by @irfansharif and @tbg, optionally @bdarnell  
- 4th commit (init scripts) for review by @bobvawter, @rafiss, @bdarnell  and @apantel 

Please peruse the commit messages and especially the release notes in the individual commits for details and motivation about the change.

### CODE REVIEW

**Comment by knz** [comment]
@celiala @aaron-crl I think this is the infrastructure needed to set up a predefined  operator role on CC clusters.

**Comment by bobvawter** [comment]
Some user-experience thoughts:
  * Is it possible to make the init scripts atomic?  That is, they must all complete successfully or not at all.
  * If an init script fails, does the node fail to start (and is then restarted?). Does it continue to flap?

I'd like to avoid the situation where a cluster only partially initializes and then requires all of the data directories to be reset in order to try again.

**Comment by jordanlewis** [comment]
cc @vy-ton  as well, since this feels related to schema management tools.

**Comment by knz** [comment]
>    Is it possible to make the init scripts atomic? That is, they must all complete successfully or not at all.

The scripts can contain BEGIN .. COMMIT and then be assembled using nested transactions, so yes. That's a task for the author of the scripts. 

(We are restricted with how our DDL behaves wrt SQL txn boundaries, but that will be worked on throughout the v21.1 cycle by the schema team, independently)

>  If an init script fails, does the node fail to start (and is then restarted?). Does it continue to flap?

The outcome I want to maintain throughout the design/iterations is that **no client connections* should be accepted by _any_ node in the cluster until all scripts have finished executing.**

Then we can decide how errors should be handled:

- currently the code causes the *first node* to stop (and thus flap) upon encountering an error.

  The other nodes remain _waiting_ (but do not accept client connections).

- we could change this to have the first node attempt to re-run the scripts (in a loop, with a wait time) instead, until all succeed

- other options are possible? I am seeking your input as to what is reasonable.

**What is the standard operating procedure with other products when an init script fails?** 

(My own instinct would be to tear down the new cluster entirely, and restart from scratch. In which case, the current behavior is not too bad. WDYT?)

**Comment by knz** [comment]
I would also suggest the following course of action:

1. merge this PR (or a minimally improved version of it, after review) as MVP
2. use the 21.1 cycle to improve upon it to give the feature the desired semantics incrementally

In particular it's going to be much easier experimenting the integration with 3rd party software stacks if we have the feature "baked in", instead of letting this PR iterate in an open state forever.

**Comment by rafiss** [comment]
> What is the standard operating procedure with other products when an init script fails?

Excerpt from docs of the Postgres docker image: https://hub.docker.com/_/postgres
```
Warning: scripts in /docker-entrypoint-initdb.d are only run if you start the container with a data directory that is empty; any pre-existing database will be left untouched on container startup. One common problem is that if one of your /docker-entrypoint-initdb.d scripts fails (which will cause the entrypoint script to exit) and your orchestrator restarts the container with the already initialized data directory, it will not continue on with your scripts.
```

That's slightly different since it's operating at a layer above what we have here, but I would still want that behavior. I.e., if the init-SQL scripts fail, then the init command should fail

**Review by rafiss — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54771)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @knz, and @tbg)

___
*[pkg/cli/initial_sql.go, line 75 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54771#-MI5JFUaE0rQCLvC9lHD:-MI5JFUaE0rQCLvC9lHE:bav1bef) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5f806456f5e496afad7fb3793de9d2ff99ccad/pkg/cli/initial_sql.go#L75)):*
> ```Go
> 
> 	// If there are initialization scripts, run them.
> 	if err := maybeRunScripts(ctx, s); err != nil {
> ```

just to check, this happens for both `init` and `start-single-node`?
___
*[pkg/cli/initial_sql.go, line 205 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54771#-MI5KCYICvNolb8RRcVV:-MI5KCYICvNolb8RRcVW:b-93zff1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5f806456f5e496afad7fb3793de9d2ff99ccad/pkg/cli/initial_sql.go#L205)):*
> ```Go
> 	}
> 
> 	scripts, err := filepath.Glob(filepath.Join(startCtx.initialSQLDir, "*.sql"))
> ```

nit: the file suffix match should be case-insensitive
___
*[pkg/cli/initial_sql.go, line 229 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/54771#-MI5JxQZ49snVvbRCSGg:-MI5JxQ_-iFPtxlGp3jY:b3dv4ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e5f806456f5e496afad7fb3793de9d2ff99ccad/pkg/cli/initial_sql.go#L229)):*
> ```Go
> 	// TODO(knz): We may want to redirect stdout/stderr
> 	// from the script to the log file.
> 	for _, script := range scripts {
> ```

is it important to sort the scripts here to make sure they are run in the same order? or is filepath.Glob already sorted?

**Comment by knz** [comment]
> if the init-SQL scripts fail, then the init command should fail

Can you explain this more. 

I understand that there needs to be a control that the scripts have executed.

Why does it matter that the control be the "init" command *specifically*?

As proposed here the control can be a simple SQL command issued after "init": if the script(s) fail, the SQL control is guaranteed to fail to connect or timeout. How is that insufficient?

**Comment by vy-ton** [comment]
> CockroachDB runs `.sql` files from the directory specified via `--initial-sql-dir` exactly once (only on 1 node)

Just to make sure I understand, only 1 node in your cluster can specify `--initial-sql-dir`?

My larger question is what are the tradeoffs of doing this with the current approach vs at a higher level (i.e Docker, Operator, or a schema migration tool running these initialization scripts). If multiple users are sharing a cluster, a user could be blocked by an error in someone else's SQL files?

**Comment by rafiss** [comment]
> Why does it matter that the control be the "init" command specifically?
>
> As proposed here the control can be a simple SQL command issued after "init": if the script(s) fail, the SQL control is guaranteed to fail to connect or timeout. How is that insufficient?

I misstated -- I didn't mean that the `init` command specifically should fail. I had just assumed that `--initial-sql-dir` was only going to be used with the `init` command. I should have said "the command" should fail.

**Comment by knz** [comment]
Vy the flag should be specified for all nodes. This is the standard way anyway, with orchestration it's desirable that the same command line be used for every pod.

Then regarding your second concern.

I've thought long and hard about using an external mechanism instead. That's what the two original issues were proposing as "workaround". Except it really, really doesn't work.

First of all, we need semantics such that regular applications can't establish a connection before the initialization completes. We simply can't implement such a global lock using regular SQL connections.

Then theres a pretty big security catch-22. A cluster starts with just 'root' defined. For security reasons we'd *strongly* prefer that the root TLS certificate never be copied outside of the VMs/pods running the cluster, as it could be misused or misplaced. The security best practice is to define an operator user with just enough privileges to set up the applications databases and tables. But then we need the root account to create that operator user. How to use the root account from a regular SQL client, "outside" of the cluster nodes, if we don't want to copy the root TLS cert around?

Finally, there are hard questions to answer in the face of operator mistakes or other spurious infra bugs, such as a node accidentally becoming unavailable during the initialization. How to ensure that the scripts have run, or not run, and get certain about the outcome using a regular SQl connection? What if it drops?



Having the mechanism privileged "inside" cockroachdb buys us these 3 things that are astonishingly hard to get using the regular SQL machinery:

- secure enclave so the root credentials never have to leak out

- guaranteed exactly-once execution

- guarantee that client apps can only start processing after the initialization has completed.



I would certainly welcome any news that these concerns do not matter and we could do away with something simpler!

**Comment by knz** [comment]
> If multiple users are sharing a cluster, a user could be blocked by an error in someone else's SQL files?

This PR is about cluster-wide initialization, when the cluster has just been created. At that point there are no multiple users yet.

(Unless I misunderstood your question and you are seeking some solution for per-tenant SQL initialization? This PR does not achieve that but the mechanism is general and can be extended in that direction. 

In any case the scripts run and must complete before the very first client connection is allowed.)

**Comment by knz** [comment]
> I had just assumed that --initial-sql-dir was only going to be used with the init command. I should have said "the command" should fail.

Yeah it's my bad for not explaining how the orchestration layer can wait on completion of the initialization. I'll work to include examples in tests and the commit message.

**Comment by vy-ton** [comment]
Is `--initial-sql-dir` a directory hosted somewhere like S3?

@knz Thank you for the explanation. As you described, using `--initial-sql-dir` for the cluster initialization user story makes sense. 

I think we need to clarify, probably with docs, how developers should be using this feature. Using `--initial-sql-dir` for some of the original suggestions such as `create some initial tables` or `load an initial dataset using IMPORT or RESTORE` I believe could be an anti-pattern for production, maybe fine for local/development/testing which is where we've seen users wanting https://github.com/cockroachdb/cockroach/issues/48647.

I'm distinguishing between cluster initialization and application initialization use cases. This feature seems great for cluster initialization which should be a smaller set of SQL which almost never fails. Whereas, application initialization involves a greater set of SQL statements (schema initialization, schema migrations, data import) and higher chance of errors. Leaving application initialization "outside" of cockroachdb I believe gives better visibility and error handling (something we can/should validate with our users).

**Comment by knz** [comment]
> Is --initial-sql-dir a directory hosted somewhere like S3?

No. This is usually a k8s "secrets" store or reusable data volume, attached to a container as volume mount.

> I'm distinguishing between cluster initialization and application initialization use cases. This feature seems great for cluster initialization which should be a smaller set of SQL which almost never fails.

Yes this is a good product-level summary of the assumptions I've made. I'll copy your sentence in the commit message.

For context, this work originally stems from multiple requests by the CC platform team.

**Comment by vy-ton** [comment]
> Warning: scripts in /docker-entrypoint-initdb.d are only run if you start the container with a data directory that is empty; any pre-existing database will be left untouched on container startup

Will CRDB mirror the above PG behavior as well?

**Comment by knz** [comment]
> Will CRDB mirror the above PG behavior as well?

The code I implemented does this for new cluster (created after my patch).
For already-existing clusters (initialized in a previous version), it will run the scripts (just once) on top of the existing cluster. I could amend the change to recognize that situation.

**Review by irfansharif — COMMENTED** [review]
> if that SQL fails the marker is not created. This gives an opportunity to the DBA to re-start the first node after fixing the initial SQL, to proceed successfully.

Noob question: if every node startup blocks on the initial SQL event being present, and if something goes wrong when running the initial SQL, upon restart, will it still succeed? (I'm likely conflating disallowing client connections with "is KV up and running"). Either way if it's not too cumbersome, I think the machinery introduced in ghis "exactly-once cluster init" commit deserves a test.

I'm also not sure what the best workflow is for when the initial SQL provided fails. In another comment you suggested tearing it all down and restarting. We have something similar in our `cockroach init` codepaths where if the bootstrap process fails once, we block out all future attempts as well. If this is the same UX we want here (I'm not sure if we do or don't), could we enforce it similarly?

https://github.com/cockroachdb/cockroach/blob/1be32272030366639be5b4093c6d45bc143978ba/pkg/server/init.go#L79-L81

---

What do we want to have happen when the operator does not provide any initial SQL when they start the cluster up for the first time? (or upgrade from a previous version for that matter). Do we still want an even to get put into `system.eventlog`? Perhaps this should also make it into the release note?


> Alternatively, they can manually insert an entry in `system.eventlog` with the column `"entryType"` set to `initial_sql_executed` *prior to the upgrade*. This entry will cause the cluster to skip over the SQL initialization script after the upgrade.

Unscientifically this seems like it should not require manual user intervention. Is it not acceptable to only have this mechanism enabled for clusters starting at v21.1+, given it's primarily concerned with "initial SQL"? Would peeking into bootstrap version help us achieve that UX?

**Review by bdarnell — COMMENTED** [review]
Making this an argument to `--start` seems like a very awkward user experience. What I had envisioned for #15301 was some initial-sql arguments to `cockroach init`. And I had envisioned those as analogous to the arguments of `cockroach sql` - the `-e` and now `-f` flags to run strings on the command-line or a single file. Doing this in `init` seems to provide a cleaner path to waiting for the SQL to finish and providing reasonable error feedback. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54771#-:-MIKaDL-E6LTkaOpBqWg:bc7dyhd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @tbg)

**Comment by vy-ton** [comment]
> For already-existing clusters (initialized in a previous version), it will run the scripts, just once, on top of the existing cluster. I could amend the change to recognize that situation.

Without this amendment, won't users have to write these SQL scripts to be idempotent because the initial roles/users would have been created in the previous version?

Why are we supporting multiple `.sql` files? Is there a guaranteed order when applying the scripts?

**Comment by knz** [comment]
Today's answer is in two parts. First I answer the questions that were not answered yet from the comments above. Then I summarize the "next steps".

### Answers

> @irfansharif if every node startup blocks on the initial SQL event being present, and if something goes wrong when running the initial SQL, upon restart, will it still succeed?

Currently, all nodes but the first wait until the magic row is inserted in `system.event_log`, upon every restart. The first node will either skip and start normally if the row already exists, or actually run the SQL init scripts. If the scripts fail, the magic row is not added, the first node stops, and the operator is given a second chance (while the other nodes remain waiting).

> @irfansharif I'm also not sure what the best workflow is for when the initial SQL provided fails. In another comment you suggested tearing it all down and restarting. We have something similar in our cockroach init codepaths where if the bootstrap process fails once, we block out all future attempts as well. If this is the same UX we want here (I'm not sure if we do or don't), could we enforce it similarly?

- Is this the same UX we want here? I don't know. I was hoping that @vy-ton @rafiss  or @bobvawter would tell us what is best wrt UX upon errors.

- In either case (either non-recoverable fail, or give-a-chance-to-the-DBA fail) it's easy to implement. I'm waiting for guidance to adjust the code accordingly.

> @irfansharif What do we want to have happen when the operator does not provide any initial SQL when they start the cluster up for the first time? [...] Do we still want an even to get put into system.eventlog? Perhaps this should also make it into the release note?

The code as-is simply treats this the same as "initial SQL was provided but is empty": the magic row is inserted in `system.event_log`, then all nodes start as normally, and no future initial SQL will ever run after that.

Yes I agree this should make its way in the release note.

> @irfansharif [regarding the suggestion to manually insert a row in `event_log` during an upgrade] Unscientifically this seems like it should not require manual user intervention. Is it not acceptable to only have this mechanism enabled for clusters starting at v21.1+, given it's primarily concerned with "initial SQL"? Would peeking into bootstrap version help us achieve that UX?

Agreed. I came to the same conclusion after other comments in this thread. 

> @vy-ton Without this amendment, won't users have to write these SQL scripts to be idempotent because the initial roles/users would have been created in the previous version but then the scripts get executed on the new version?

Yes that is correct. Based on @irfansharif's other comment, I think it would be good to amend the code anyway, because if the cluster has already been initialized it would be quite the UX surprise to see initialization scripts run again.

> @vy-ton If we mirror the PG behavior, it would not support a user amending their cluster initialization. Its unclear to me whether users want a cluster initialization that happens only once ever or that occurs on every cluster restart.

I have thought a little bit about this. To me, these are two separate use cases:

- just-once cluster initialization

  For this we have clear user demand: set up initial users/roles, set up reference tables, import data etc.

- additional SQL logic to run upon node restarts (the phrase "cluster restart" has no meaning in cockroachdb -- there is no difference between restarting nodes one by one or all at once)

  For this I don't know we have user demand. We could envision users who want to maintain their own SQL-level statistics about cluster health. Maybe? It is not too complicated to add a separate directory with scripts that would run upon node restarts.

> @vy-ton Why are we supporting multiple .sql files?

I have learned (see #48647) that this is the idiom for folk running databases inside containers. It makes it possible to combine initialization scripts for multiple applications / dev teams more easily.

> @vy-ton Is there a guaranteed order when applying the scripts?

Yes. @rafiss was asking the same: we use a file listing that sorts the file names alphanumerically.
The idiom is to create SQL script names prefixed by a number, like `10-init-users.sql`, `20-create-dbs.sql` etc.

> @bdarnell What I had envisioned for #15301 was some initial-sql arguments to cockroach init. I had envisioned those as analogous to the arguments of cockroach sql - the `-e` and now `-f` flags to run strings on the command-line or a single file.

Yes I remember this and I had envisioned the same. It's not even hard to implement.

However as I learned since, the idiom for container-based deployments is to assemble initialization bits from multiple teams, separately from the (ops) team who creates the orchestration scripts. It's awkward to have to edit a (long) command-line every time a team wants to add or remove a bit of SQL initialization.

Additionally, a volume mount makes it easier to import directories of SQL scripts from source control in tests, during development iterations etc.

Finally, there's a security argument. The init SQL needs to run with root credentials: at that point there are no other users defined. However, we have a strong security mandate to avoid leaking the root TLS cert outside of the container pods -- we want to apply the principle of least privileges. The `init` command does not need more privileges than just calling the `Bootstrap` function.

> @bdarnell  Doing this in init seems to provide a cleaner path to waiting for the SQL to finish and providing reasonable error feedback.

I agree with you on the UX end result: the `cockroach init`  **command** should probably wait on SQL initialization. But I disagree on whether the `Bootstrap` RPC should be responsible for this. As we iterate on this PR, I plan to extend the code for `cockroach init` to subsequently open a SQL connection and poll until the `event_log` row has been create.

### Next steps

- [ ] iterate with @vy-ton @rafiss @bdarnell @bobvawter on the desired UX when there are errors in the SQL scripts:
  - Should the initialization be considered as irremediably failed?
  - Should the DBA be given a chance to fix the scripts and retry?
  - Should all the data created by scripts so far be wiped upon a SQL init failure, so that the next run will see an empty cluster again?
  - Should the `cockroach init` CLI tool wait until the SQL initialization completes?

- [ ] assuming there's agreement on whether `cockroach init` should wait on initialization, modify the `init` code accordingly.
- [ ] add more tests for the exactly-once initialization as suggested by @irfansharif 
- [ ] add more tests to verify the order of the SQL initialization (we're assuming the order because that's what `filepath.Glob` does, but it's not checked in unit tests) as suggested by @rafiss 
- [ ] extend the commit messages and PR description with a FAQ of all user-facing concerns that were raised during the review.

**Comment by knz** [comment]
I would also like to thank everyone who participated so far.

It helped us collectively understand that the problem statement (in  #15301 then  #48647) was incomplete.

I'm still gathering comments on this PR -- especially if you have follow-ups to the point already raised. But at some point in the near future I also propose to lift the discussion into an RFC.

**Comment by bdarnell** [comment]
> I have learned (see #48647) that this is the idiom for folk running databases inside containers. It makes it possible to combine initialization scripts for multiple applications / dev teams more easily.

Interesting. That makes the directory approach make more sense. However, I still feel like it makes more sense to mount this directory on the (single) init command/container instead of the (many) node containers. 

> The init command does not need more privileges than just calling the Bootstrap function.

OK, that's a good point - changes to the init command authentication might make it harder to do this securely through the init command. But I'm not convinced it would completely break this - given the inherent danger of `init` I think it's always going to need reasonably strong authentication. Do we need to defer this change until we have the TLS redesign more well-defined? Remember that one of the tentative goals of that redesign is to either create a non-root user or set a root user's password as a part of `init`. 

Alternately, we could introduce this flag as an experimental feature for `start-single-node` (which satisfies the development docker case) and later decide whether it belongs on `start` or `init` for multi-node clusters. 

Finally, remember that for the docker case in other databases, postgres et all *do not* have an analogous feature - this is all done at arms length in the scripting of the docker container, not the DB itself.

**Comment by rafiss** [comment]
> Alternately, we could introduce this flag as an experimental feature for `start-single-node` (which satisfies the development docker case) and later decide whether it belongs on `start` or `init` for multi-node clusters.
> 
> Finally, remember that for the docker case in other databases, postgres et all do not have an analogous feature - this is all done at arms length in the scripting of the docker container, not the DB itself.

I would also note that if we decide to only support this sort of initialization for `start-single-node` with no plans for allowing this for multi-node setups, then we should keep in mind that the only motivation here is to support local/CI development with Docker. In which case, _all_ of the work could be done in our Docker container script (as Postgres does).

**Comment by knz** [comment]
> @bdarnell  remember that for the docker case in other databases, postgres et all do not have an analogous feature - this is all done at arms length in the scripting of the docker container, not the DB itself.

I remember this. Rafi was also proposing to do this back in https://github.com/cockroachdb/cockroach/issues/48647#issuecomment-679370828

However when I tried that I discovered it didn't work - and I explained this above to Vy in https://github.com/cockroachdb/cockroach/pull/54771#issuecomment-699159752.

> @bdarnell Do we need to defer this change until we have the TLS redesign more well-defined? 

I wasn't planning on completing the work until a bit later in the release cycle. I started the PR early so as to gather feedback and additional requirements (it worked very well for this purpose!) and plan the remaining work ahead.

> Remember that one of the tentative goals of that redesign is to either create a non-root user or set a root user's password as a part of init.

Yes! As well as setting up the CC operator role and other requirements by SREs (possibly even creating a CC-monitoring tenant in multi-tenant crdb).

That in turn says that it's not going to be easy working around the catch-22 which I explained in https://github.com/cockroachdb/cockroach/pull/54771#issuecomment-699159752. This is why I still think this logic is best served by an authn-less loopback connection, and have the scripts run server-side.

But I propose that we explore this spectrum of design points in the RFC.

**Comment by knz** [comment]
>  we should keep in mind that the only motivation here is to support local/CI development with Docker. 

That may be the only requirement for the AppDev team, but there are other candidate users already identified.

> In which case, all of the work could be done in our Docker container script (as Postgres does).

Even for the local / `start-single-node` case, I don't think the work can be done from the script. This is explained in https://github.com/cockroachdb/cockroach/pull/54771#issuecomment-699159752 : a script does not enable you to block the connections from regular clients until the initialization completes. We don't offer a synchronization mechanism to "wait until initialization completes" with our existing APIs and interfaces. This PR here achieves that.

**Comment by knz** [comment]
Discussed with @rafiss and @ZhouXing19 today: this PR is too complex for the envisioned goal.
See discussion here: https://github.com/cockroachdb/cockroach/issues/48647#issuecomment-925147362
